### PR TITLE
Feature: Templated operations inside of interfaces

### LINKED ISSUE #1389: Implementation: Operation temaplate inside of operation
from: #1250 

```cadl
interface OpFactory<T> {
  none(): T;
  a<A>(u: A): T;
  b<B>(u: B): T;
}
```

Changes: 
- Allow operation inside of interface
- No shadowing of template params
```cadl
interface Foo<T> {
  bar<T>(): T;
  //  ^-- error shadowing T in Foo
}
```
- write docs and recommened usage of interface factory
- describe breaking change of `tempalateArguments` in api

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix #1389 

[Playground example](https://cadlplayground.z22.web.core.windows.net/prs/1479/?c=aW1wb3J0ICJAY2FkbC1sYW5nL3Jlc3QiOwoKdXNpbmcgQ2FkbC5IdHRwOwoKaW50ZXJmYWNlIFJlYWRXcml0ZTxUPiB7CiAgcmVhZCgpOiBUOwogIHfFGlI%2BKHQ6IFQpOiBSOwp9CgphbGlhcyBNeck8ID3LSHN0cmluZz47CgoKQHJvdXRlKCJteS3EVyIpCm9wIG3FOCBpzUIuxCA7zTLFfMgzxmfPNMUiPGludDMyPjsK)

Todo: 
- [x] Write breaking change or deprecation entry about templateArugments

### CODE REVIEW

File: packages/compiler/core/checker.ts [hunk 1]
```diff
@@ -935,44 +955,24 @@ export function createChecker(program: Program): Checker {
...
-              : sym.flags & SymbolFlags.Scalar
-              ? checkScalar(decl as ScalarStatementNode, mapper)
-              : sym.flags & SymbolFlags.Alias
-              ? checkAlias(decl as AliasStatementNode, mapper)
-              : sym.flags & SymbolFlags.Interface
-              ? checkInterface(decl as InterfaceStatementNode, mapper)
-              : sym.flags & SymbolFlags.Operation
-              ? checkOperation(decl as OperationStatementNode, mapper)
-              : checkUnion(decl as UnionStatementNode, mapper);
+          baseType = checkDeclaredType(decl, sym, mapper);
```
Comment (timotheeguerin) [inline packages/compiler/core/checker.ts:958]: For review: I just refactored this into its own function as we had that reused exactly the same right below

File: packages/compiler/core/checker.ts [hunk 2]
```diff
@@ -935,44 +955,24 @@ export function createChecker(program: Program): Checker {
         if (symbolLinks.declaredType) {
           baseType = symbolLinks.declaredType;
         } else {
-          baseType =
```
Comment (timotheeguerin) [inline packages/compiler/core/checker.ts:938]: For review: I just refactored this into its own function as we had that reused exactly the same right below

File: packages/compiler/core/types.ts [hunk 3]
```diff
@@ -50,7 +50,18 @@ export interface DecoratedType {
...
+  args: readonly Type[];
+  /** @internal */ map: Map<TemplateParameter, Type>;
+}
+
 export interface TemplatedTypeBase {
+  templateMapper?: TypeMapper;
+  /**
+   * @deprecated use templateMapper instead.
+   */
   templateArguments?: Type[];
```
Comment (timotheeguerin) [inline packages/compiler/core/types.ts:65]: Any opinions on what we should do here about that list. Deprecate it or just remove it?

I guess it shouldn't affect emitter and libraries that don't  see templated operation in interface if we keep it deprecated.

File: docs/language-basics/interfaces.md [hunk 4]
```diff
@@ -60,3 +60,43 @@ interface ReadWrite<T> {
   write(t: T): void;
 }
 ```
+
+## Interface operation templates
+
+Operations defined inside of interface can also be templated. ([see templates](./templates.md) for details on templates.)
```
Comment (daviwil) [inline docs/language-basics/interfaces.md:66]:
```suggestion
Operations defined inside of an interface can also be templated. ([see templates](./templates.md) for details on templates.)
```

File: docs/language-basics/interfaces.md [hunk 5]
```diff
@@ -60,3 +60,43 @@ interface ReadWrite<T> {
...
+  write<R>(t: T): R;
+}
+
+alias MyReadWrite = ReadWrite<string>;
+
+op myWrite is MyReadWrite.write<int32>;
+```
+
+:::caution
+Interface having some operation templated and some not will result in the non templated operation being picked up as operation for the service.
```
Comment (daviwil) [inline docs/language-basics/interfaces.md:80]:
```suggestion
Any uninstantiated, templated operation defined in an interface will be excluded from the list of service operations.
```

File: docs/language-basics/interfaces.md [hunk 6]
```diff
@@ -60,3 +60,43 @@ interface ReadWrite<T> {
...
+
+This applies as well if using `extends` on a templated interface that has only some templated operation.
```
Comment (daviwil) [inline docs/language-basics/interfaces.md:82]:
```suggestion
This also applies when using `extends` on an interface that contains templated operations with unfilled template arguments.
```

File: packages/compiler/core/types.ts (hunk 3 above)
Comment (daviwil) [reply packages/compiler/core/types.ts:65]: Looks like it's pretty easy to adapt existing uses of `templateArguments` in linter and decorators which check if a type is templated (check `type.templateMapper?.partial` instead) so I'm OK with deprecating it.

**Review by daviwil — APPROVED** [review]
Looks great to me, thanks a lot for getting this in!

File: packages/compiler/core/types.ts (hunk 3 above)
Comment (bterlson) [reply packages/compiler/core/types.ts:65]: Should be using `isTemplateInstance` anyway right? Good time to upgrade.

My feeling though is that we should leave this in but deprecated for a while. I fear our Morgan Stanley friends will be using this. I wonder if we could turn it into a getter that logs a deprecation warning, even?

**Review by bterlson — APPROVED** [review]

**Pull request merged**